README.md
# WpNonces README

## Description
By developing WordPress Plugins and Themes sometimes you need store your own Project-Settings into 
the wp_options table.
This Tool make it easier to handle the wp_nonce_* functions in a OOP-environment.

## Installation
Easiest way to install is using composer, 
but at moment is it a private Bitbucket Repository, and is not possible to use composer. 

Please clone this Repo with

`git clone https://your-account@bitbucket.org/pcservice-voll/wpnonces.git`

or download it from

[private Bitbucket Repo](https://bitbucket.org/pcservice-voll/wpnonces/)

## How to use

In general, you can use this class without further adjustment.

As an example I have created a plugin, where i will use this package to handle nonce.


_Step 1 :_ require the autoload.php to make it accessible in your plugin/theme


- download the Repo and store it in vendor folder in your plugin<br>
<code>
require_once WP_PLUGIN_DIR . '/' . plugin_basename( __DIR__ ) . '/vendor/WpNonces/vendor/autoload.php';
</code> 


_Step 2 :_ load an instance in your settings or options page

<code>$wp_nonces = \MediaStoreNet\WpNonces\WpNonces::getInstance();</code><br>
<b>this static method allows you to use allways the same instance of the class<br>
in all your settings/options files.</b>


_Step 3:_ create a form
In my case i do this on a separate function and call these in my settings/options page

`<?php
function my_form( $wp_nonces ) {
	ob_start();
?>`

    <form method="get" action="options.php">
        <label for="testinput">Input</label>
        <input type="text" id="testinput" name="testinput"/>
		<?php $wp_nonces->getNonceField(); ?>
		<?php submit_button( 'speichern' ); ?>
    </form>
`<?php 
	return ob_get_clean();
?>`

_Step 4:_ Validate

to validate you have several options

call **$wp_nonce->verify('nonceString')**

`if ( $wp_nonces->verifyNonce($_REQUEST['_wpnonce']) ):
    //store options...;
endif;`


call **$wp_nonce->verifyAdmin()**

`if ( $wp_nonces->verifyAdmin() ):
   // store options...;
endif;`


call **$wp_nonces->verifyAjax()** to validate Ajax Requests

`if ( $wp_nonces->verifyAjax() ):
    //store ajax request options...;
endif;`


By default is used fieldName of nonce "_wpnonce" like in WordPress use.
The action string is "wp-oop-nonce"


For more secure and customized nonce you can modify the fieldName and Action string too

`$wp_nonces->setFieldName('my-custom-name');
$wp_nonces->setAction('my-custom-action');`


When needed you can also use more then one instance like this

`$nonces1 = new MediaStoreNet\WpNonces\WpNonces();
$nonces1->setFieldName('name1');
$nonces1->setAction('action1');`
	
`$nonces2 = new MediaStoreNet\WpNonces\WpNonces();
$nonces2->setFieldName('name2');
$nonces2->setAction('action2');`



To see all available propertys and methods, please visit our 
[API Documentation Site](http://wpnonces.docs.media-store.net/)


## Minimum Requirements / Dependencies
* PHP ^7.0
* WordPress latest-2

When installed for development, via Composer requires:

* phpunit/phpunit (BSD-3-Clause)
* brain/monkey (MIT)
* inpsyde/php-coding-standards

## Documentation
**Please visit our Documentation Site**
[API Documentation Site](http://wpnonces.docs.media-store.net/)

## CHANGELOG
[Link to changelog](CHANGELOG.md)
## Licence and Copyright

GPLv2+ Licence

Copyright (c) 2019 Media-Store.net


CHANGELOG.md
# WpNonces CHANGELOG

## Versions

= v0.0.1 =

Initial Commit including
+ /src/NonceInterface.php
+ /src/WpNonces.php
+ /tests/NonceExceptionsTest
+ /tests/WpNoncesTest
+ composer.json file
+ .gitignore
+ phpunit configuration

= v0.0.2 =

- method for wp_nonce_field
- method for check_admin_referer
- method for check_ajax_referer
- tests for all this methods
- UPDATE: generated phpdoc 

= v0.0.3 =
- method for modify nonce_lifetime
- tests for nonce_lifetime
- UPDATE: generated phpdoc 


tests/WpNoncesTest.php
<?php
/**
 * Created by Media-Store.net
 * User: Artur
 * Date: 01.10.2019
 * Time: 21:17
 * PHP Version: ^7.1
 *
 * @category Wp_Nonces
 * @package  MediaStoreNet\WpNonce\Test
 * @author   Artur Voll <info@pcservice-voll.de>
 * @license  [GPLv2+] <https://www.gnu.org/licenses/old-licenses/gpl-2.0.html>
 * @link     Media-Store.net <https://media-store.net>
 */

namespace MediaStoreNet\WpNonces\Test;

use Brain\Monkey;
use MediaStoreNet\WpNonces\WpNonces;
use MonkeryTestCase\BrainMonkeyWpTestCase;

/**
 * Class WpNoncesTest
 *
 * @category Wp_Nonces
 * @package  MediaStoreNet\WpNonce\Test
 * @author   Artur Voll <info@pcservice-voll.de>
 * @license  [GPLv2+] <https://www.gnu.org/licenses/old-licenses/gpl-2.0.html>
 * @link     Media-Store.net <https://media-store.net>
 */
class WpNoncesTest extends BrainMonkeyWpTestCase
{

    /**
     * Instance of WpNonces Class
     *
     * @var WpNonces
     */
    public $nonceInstance;

    /**
     * SetUp...
     *
     * @return void
     */
    protected function setUp(): void
    {
        parent::setUp();
        Monkey\setUp();

        // define wp_functions fakes
        Monkey\Functions\stubs(
            [
                'wp_create_nonce'     => function ($action) {
                    return md5($action);
                },
                'wp_nonce_url'        => function ($actionurl, $nonce, $name) {
                    return sprintf(
                        '%s?%s=%s',
                        $actionurl,
                        $name,
                        $nonce
                    );
                },
                'wp_nonce_field'      => function (
                    $action,
                    $name,
                    $referer = true,
                    $echo = true
                ) {
                    $field1 = sprintf(
                        '<input type="hidden" %1$s="%2$s" />',
                        $name,
                        $action
                    );
                    $field2 = sprintf(
                        '<input type="hidden" %1$s="%2$s" />'
                    );

                    $output = $field1;
                    if ($referer) :
                        $output .= '<br>' . $field2;
                    endif;

                    if ($echo) :
                        echo $output;
                    else:
                        return $output;
                    endif;
                },
                'wp_verify_nonce'     => function ($nonce, $action) {
                    if (md5($action) === $nonce) {
                        return 1;
                    } else {
                        return false;
                    }
                },
                'check_admin_referer' => function ($action, $query_arg) {
                    if (isset($_REQUEST[$query_arg])
                        && $_REQUEST[$query_arg] === md5($action)
                    ) {
                        return true;
                    } else {
                        return 'not';
                    }
                },
                'check_ajax_referer'  => function ($action, $query_arg, $die) {
                    if (isset($_REQUEST[$query_arg])
                        && $_REQUEST[$query_arg] === md5($action)
                    ) {
                        return true;
                    } else {
                        return $die ? 'not' : false;
                    }
                }
            ]
        );

        $this->nonceInstance = WpNonces::getInstance();
    }

    /**
     * TearDown...
     *
     * @return void
     */
    protected function tearDown(): void
    {
        parent::tearDown();
        Monkey\tearDown();
        unset($this->nonceInstance);
    }

    /**
     * Testing static getInstances method
     *
     * @return string
     */
    public function testGetInstance()
    {
        self::assertInstanceOf(
            WpNonces::class,
            $this->nonceInstance,
            'Expected a Instance of WpNonces Class'
        );

    }

    /**
     * Testing give an correct NonceString
     *
     * @return string
     * @throws \Exception
     */
    public function testGetNonce()
    {

        $this->nonceInstance->setAction('wp-oop-nonce');

        self::assertSame(
            md5($this->nonceInstance->getAction()),
            $this->nonceInstance->getNonce(),
            'Expected a genareted md5 string of "action" '
        );
    }

    /**
     * Testing get an correct action string
     *
     * @return string
     */
    public function testGetAction()
    {
        self::assertSame(
            'wp-oop-nonce',
            $this->nonceInstance->getAction(),
            'Expected a String of "wp-oop-nonce" '
        );
    }

    /**
     * Testing get an correct fieldName
     *
     * @return string
     */
    public function testGetFieldName()
    {
        self::assertSame(
            '_wpnonce',
            $this->nonceInstance->getFieldName(),
            'Expected a String of "_wpnonce" '
        );
    }

    /**
     * Testing get an correct NonceUrl String
     *
     * @return string
     */
    public function testGetNonceUrl()
    {
        $url      = $this->nonceInstance->getNonceUrl('http://test.de');
        $expected = 'http://test.de?' .
                    $this->nonceInstance->getFieldName() . '=' .
                    $this->nonceInstance->getNonce();

        self::assertSame(
            $expected,
            $url,
            'Expected a string of url with integrated nonce'
        );
    }

    /**
     * Testing set the action to string test
     *
     * @return string
     * @throws \Exception
     */
    public function testSetAction()
    {
        $this->nonceInstance->setAction('test');

        self::assertSame(
            'test',
            $this->nonceInstance->getAction(),
            'Expected a string of test'
        );

    }


    /**
     * Testing set the fieldName to my-nonce string
     *
     * @return string
     */
    public function testSetFieldName()
    {
        $this->nonceInstance->setFieldName('my-nonce');

        self::assertSame(
            'my-nonce',
            $this->nonceInstance->getFieldName(),
            'Expected a string of my-nonce'
        );

    }

    /**
     * Testing verifying of Nonce String
     *
     * @return string
     */
    public function testVerifyNonce()
    {
        self::assertEquals(
            1,
            $this->nonceInstance->verifyNonce($this->nonceInstance->getNonce()),
            'Expected a true condition'
        );
    }

    /**
     * Testing verifying check_admin_referer
     *
     * @return string
     */
    public function testVerifyAdmin()
    {
        $nonceName = $this->nonceInstance->getFieldName();
        // set $_REQUEST
        $_REQUEST[$nonceName] = $this->nonceInstance->getNonce();

        self::assertTrue(
            isset($_REQUEST[$nonceName]),
            'Expected that the $_REQUEST of field_name is setted'
        );

        self::assertEquals(
            md5($this->nonceInstance->getAction()),
            $_REQUEST[$nonceName],
            'Expected that the $_REQUEST nonce is the same as md5() of action'
        );

        self::assertTrue(
            $this->nonceInstance->verifyAdmin(),
            'Expected that verifyAdmin is truely'
        );

        // set $_REQUEST to wrong nonce
        $_REQUEST[$nonceName] = 'abcde';
        self::assertSame(
            'not',
            $this->nonceInstance->verifyAdmin(),
            'Expected is a die() with "not" string'
        );
    }

    /**
     * Testing verifying Ajax Request
     *
     * @return string
     */
    public function testVerifyAjax()
    {
        $nonceName = $this->nonceInstance->getFieldName();
        // set $_REQUEST
        $_REQUEST[$nonceName] = $this->nonceInstance->getNonce();

        self::assertTrue(
            isset($_REQUEST[$nonceName]),
            'Expected that the $_REQUEST of field_name is setted'
        );

        self::assertEquals(
            md5($this->nonceInstance->getAction()),
            $_REQUEST[$nonceName],
            'Expected that the $_REQUEST nonce is the same as md5() of action'
        );

        self::assertTrue(
            $this->nonceInstance->verifyAjax(),
            'Expected that verifyAdmin is truely'
        );

        // set $_REQUEST to wrong nonce
        $_REQUEST[$nonceName] = 'abcde';
        self::assertSame(
            'not',
            $this->nonceInstance->verifyAjax(),
            'Expected is a die() with "not" string'
        );
    }

    /**
     * Testing setNonceLife filter
     *
     * @return string
     */
    public function testSetNonceLifetime()
    {
        define('HOUR_IN_SECONDS', intval(3600));

        $this->nonceInstance->setNonceLifetime();

        self::assertFalse(
            has_filter(
                'nonce_life',
                [$this->nonceInstance, 'setNonceLifetime']
            ),
            'Expected that a filter not fired to nonce_life'
        );

        $this->nonceInstance->setNonceLifetime(6);

        self::assertTrue(
            has_filter(
                'nonce_life',
                ['MediaStoreNet\WpNonces\WpNonces', 'setNonceLifetime']
            ),
            'Expected that a filter fired to nonce_life'
        );
    }

}


tests/NoncesExceptionsTest.php
<?php
/**
 * Created by Media-Store.net
 * User: Artur
 * Date: 01.10.2019
 * Time: 21:36
 * PHP Version: ^7.1
 *
 * @category Wp_Nonces_Tests
 * @package  MediaStoreNet\WpNonce\Test
 * @author   Artur Voll <info@pcservice-voll.de>
 * @license  [GPLv2+] <https://www.gnu.org/licenses/old-licenses/gpl-2.0.html>
 * @link     Media-Store.net <https://media-store.net>
 */

namespace MediaStoreNet\WpNonces\Test;

use Brain\Monkey;
use MediaStoreNet\WpNonces\WpNonces;
use MonkeryTestCase\BrainMonkeyWpTestCase;

/**
 * Class NoncesExceptionsTest
 *
 * @category Wp_Nonces_Tests
 * @package  MediaStoreNet\WpNonce\Test
 * @author   Artur Voll <info@pcservice-voll.de>
 * @license  [GPLv2+] <https://www.gnu.org/licenses/old-licenses/gpl-2.0.html>
 * @link     Media-Store.net <https://media-store.net>
 */
class NoncesExceptionsTest extends BrainMonkeyWpTestCase
{

    /**
     * Representate a Instance of WpNonces Class
     *
     * @var WpNonces
     */
    public $nonceInstannce;

    /**
     * SetUp...
     *
     * @return void
     */
    protected function setUp(): void
    {
        parent::setUp();
        Monkey\setUp();
    }

    /**
     * TearDown...
     *
     * @return void
     */
    protected function tearDown(): void
    {
        parent::tearDown();
        Monkey\tearDown();
        unset($this->nonceInstannce);
    }

    /**
     * Testing if the WordPress Functions doesn't exists
     * get a Exception to user
     *
     * @return void
     */
    public function testInstanceInitException()
    {
        Monkey\Functions\stubs(
            [
                '__'       => function ($m, $l) {
                    return $m;
                },
                'esc_attr' => function ($m) {
                    return $m;
                },
                'esc_html' => function ($m) {
                    return $m;
                }
            ]
        );

        try {
            $this->nonceInstannce = WpNonces::getInstance();
        } catch ( \Exception $exception ) {
            self::assertInstanceOf(
                \Exception::class,
                $exception,
                'Expected a Exception, because the function "add_action" is not exists'
            );
        }
    }

    /**
     * Testing if try create the instance on not a options or settings page
     * This test fire a action on admin_notices
     *
     * @return void
     */
    public function testInstanceInitAdminNotice()
    {

        Monkey\Functions\stubs(
            [
                '__'       => function ($m, $l) {
                    return $m;
                },
                'esc_attr' => function ($m) {
                    return $m;
                },
                'esc_html' => function ($m) {
                    return $m;
                }
            ]
        );

        try {
            $this->nonceInstannce = WpNonces::getInstance();
        } catch ( \Exception $exception ) {
            self::assertTrue(
                has_action('admin_notices', $this->nonceInstannce),
                'Expected a action hook on admin notices'
            );
        }
    }

    /**
     * Testing if the Instance is succefull created
     *
     * @return void
     */
    public function testInstanceInitSucces()
    {
        $this->getNonceInstance();

        self::assertTrue(
            md5($this->nonceInstannce->getAction())
            ===
            $this->nonceInstannce->getNonce(),
            'Expected a String of nonce'
        );

    }

    /**
     * Create a dummy instance and set it into
     * $this->nonceInstannce
     *
     * @return void
     */
    public function getNonceInstance()
    {
        Monkey\Functions\stubs(
            [
                'add_action'          => true,
                'wp_create_nonce'     => function ($action) {
                    return md5($action);
                },
                'wp_nonce_url'        => true,
                'wp_nonce_field'      => true,
                'wp_verify_nonce'     => true,
                'check_admin_referer' => true,
                'check_ajax_referer'  => true
            ]
        );

        $this->nonceInstannce = WpNonces::getInstance();
    }

}


src/NonceInterface.php
<?php
/**
 * Created by Media-Store.net
 * Date: 30.09.2019
 * Time: 19:04
 * PHP Version: ^7.0
 *
 * @category Wp_Nonces
 * @package  MediaStoreNet\WpNonce
 * @author   Artur Voll <info@pcservice-voll.de>
 * @license  [GPLv2+] <https://www.gnu.org/licenses/old-licenses/gpl-2.0.html>
 * @link     Media-Store.net <https://media-store.net>
 */

namespace MediaStoreNet\WpNonces;


/**
 * Interface NonceInterface
 *
 * @category Wp_Nonces
 * @package  MediaStoreNet\WpNonce
 * @author   Artur Voll <info@pcservice-voll.de>
 * @license  [GPLv2+] <https://www.gnu.org/licenses/old-licenses/gpl-2.0.html>
 * @link     Media-Store.net <https://media-store.net>
 */
interface NonceInterface
{

    /**
     * Getter for the FieldName
     *
     * @return mixed
     */
    function getFieldName();

    /**
     * Setter for FieldName
     *
     * @param string $field_name // Name of the field
     *
     * @return mixed
     */
    function setFieldName(string $field_name);

    /**
     * Getter for Action
     *
     * @return mixed
     */
    function getAction();

    /**
     * Setter for Action
     *
     * @param string $action //
     *
     * @return mixed
     */
    function setAction(string $action);

    /**
     * Getter for Nonce
     *
     * @return mixed
     */
    function getNonce();

    /**
     * Getter for URL
     *
     * @param string $actionurl //
     *
     * @return mixed
     */
    function getNonceUrl(string $actionurl);

    /**
     * Returns hidden input fileds for Nonce and Referrer(optional)
     * By Default would echo the fields, by setting the $echo to false
     * the fields will be returned
     *
     * @param string $action  // Action String
     * @param string $name    // Name String
     * @param bool   $referer // show the referrer-field
     *                        // by default is true
     * @param bool   $echo    // echo or return the fields
     *                        // by default will be echo
     *
     * @return string
     */
    function getNonceField(string $action, string $name, bool $referer, bool $echo);

    /**
     * Verifyer for Nonce
     *
     * @param string $nonce //
     *
     * @return mixed
     */
    function verifyNonce(string $nonce);

    /**
     * Verifyer to Check Admin
     *
     * @return mixed
     */
    function verifyAdmin();

    /**
     * Verifyer to Check Ajax
     *
     * @return mixed
     */
    function verifyAjax();

}


src/WpNonces.php
<?php
/**
 * Created by Media-Store.net
 * Date: 30.09.2019
 * Time: 19:01
 * PHP Version: ^7.0
 *
 * @category Wp_Nonces
 * @package  MediaStoreNet\WpNonce
 * @author   Artur Voll <info@pcservice-voll.de>
 * @license  [GPLv2+] <https://www.gnu.org/licenses/old-licenses/gpl-2.0.html>
 * @link     Media-Store.net <https://media-store.net>
 */

namespace MediaStoreNet\WpNonces;

/**
 * Class WpNonces
 *
 * @category Wp_Nonces
 * @package  MediaStoreNet\WpNonce
 * @author   Artur Voll <info@pcservice-voll.de>
 * @license  [GPLv2+] <https://www.gnu.org/licenses/old-licenses/gpl-2.0.html>
 * @link     Media-Store.net <https://media-store.net>
 */
class WpNonces implements NonceInterface
{

    /**
     * Static Instance of this Class
     *
     * @var $_instance //Static Instance of this Class
     */
    private static $_instance;

    /**
     * Default FieldName
     *
     * @var string $field_name // By default is the field_name "_wpnonce"
     */
    protected $field_name = '_wpnonce';

    /**
     * Default Action for Nonce
     *
     * @var string $action // By default is action "wp-oop-nonce"
     */
    protected $action = 'wp-oop-nonce';

    /**
     * Generated Nonce String
     *
     * @var string $nonce //Generated Nonce String
     */
    protected $nonce;

    /**
     * Allows to modify default lifetime of the Nonce
     *
     * @var int $lifetime
     */
    protected $lifetime;

    /**
     * GetInstance is a static function to return of this Class Instance
     *
     * @return WpNonces
     */
    public static function getInstance()
    {
        // Initialize the service if it's not already set.
        if (self::$_instance === null) {
            self::$_instance = new WpNonces();
        }

        // Return the instance.
        return self::$_instance;
    }

    /**
     * WpNonces constructor.
     */
    public function __construct()
    {
        $this->init();
    }


    /**
     * Getter for FieldName
     *
     * @return string
     */
    public function getFieldName(): string
    {
        return $this->field_name;
    }


    /**
     * Setter for FieldName
     *
     * @param string $field_name //
     *
     * @return void
     */
    public function setFieldName(string $field_name)
    {
        $this->field_name = $field_name;
    }


    /**
     * Getter for Nonce String
     *
     * @return string
     */
    public function getNonce(): string
    {
        return $this->nonce;
    }

    /**
     * Getter for Action
     *
     * @return string
     */
    public function getAction(): string
    {
        return $this->action;
    }

    /**
     * Setter for Action
     *
     * @param string $action //
     *
     * @return void
     * @throws \Exception
     */
    public function setAction(string $action)
    {
        if ($action !== $this->action) :
            $this->action = $action;
            $this->init();
        endif;
    }

    /**
     * Getter for NonceUrl String
     *
     * @param string $actionurl //
     *
     * @see    https://codex.wordpress.org/Function_Reference/wp_nonce_url
     * @return string
     */
    public function getNonceUrl(string $actionurl): string
    {
        return wp_nonce_url($actionurl, $this->nonce, $this->field_name);
    }

    /**
     * Returns hidden input fileds for Nonce and Referrer(optional)
     * By Default would echo the fields, by setting the $echo to false
     * the fields will be returned
     *
     * @param string $action  // Action String
     * @param string $name    // Name String
     * @param bool   $referer // show the referrer-field
     *                        // by default is true
     * @param bool   $echo    // echo or return the fields
     *                        // by default will be echo
     *
     * @return string
     * @throws \Exception
     * @see    https://codex.wordpress.org/Function_Reference/wp_nonce_field
     */
    public function getNonceField(
        string $action = '',
        string $name = '',
        bool $referer = true,
        bool $echo = true
    ) {
        if (!empty($name)) :
            $this->setFieldName($name);
        endif;
        if (!empty($action)) :
            $this->setAction($action);
        endif;

        return wp_nonce_field($this->action, $this->field_name, $referer, $echo);
    }


    /**
     * Verifyer for the Nonce String
     *
     * @param string $nonce // given nonce-string
     *
     * @see    https://codex.wordpress.org/Function_Reference/wp_verify_nonce
     * @return int|bool
     */
    public function verifyNonce(string $nonce): bool
    {
        return wp_verify_nonce($nonce, $this->action);
    }

    /**
     * Verifyer for Admin Area
     *
     * @see    https://codex.wordpress.org/Function_Reference/check_admin_referer
     * @return mixed|void
     */
    public function verifyAdmin()
    {
        return check_admin_referer($this->action, $this->field_name);
    }

    /**
     * Verifyer for Ajax Requests
     *
     * @see    https://codex.wordpress.org/check_ajax_referer
     * @return mixed|void
     */
    public function verifyAjax()
    {
        return check_ajax_referer($this->action, $this->field_name, true);
    }


    /**
     * Sets filter to custom Nonce Lifetime
     * by default is lifetime 24 hours
     * You will find more Info in the WordPress Codex - Modifying the nonce system
     *
     * @param int $hours // Time in hours
     *
     * @see    https://codex.wordpress.org/WordPress_Nonces
     * @return void
     */
    public function setNonceLifetime(int $hours = 0)
    {
        if ($hours <= 0) :
            return;
        endif;

        $this->lifetime = intval($hours) * HOUR_IN_SECONDS;

        add_filter(
            'nonce_life',
            function () {
                return $this->lifetime;
            }
        );
    }

    /**
     * Initialisation / create of the Nonce
     *
     * @see    https://codex.wordpress.org/Function_Reference/wp_create_nonce
     * @return string|void
     * @throws \Exception
     */
    protected function init()
    {
        // make sure that's WordPress
        $wp_functions = [
            'wp_create_nonce',
            'wp_nonce_url',
            'wp_nonce_field',
            'wp_verify_nonce',
            'check_admin_referer',
            'check_ajax_referer'
        ];

        // return a Exception if not
        if (!function_exists('add_action')) :

            return $this->throwError(
                \BadMethodCallException::class,
                sprintf(
                    'The function "%s" is not defined, make sure you use this package on WordPress',
                    'add_action'
                )
            );


        elseif (function_exists('add_action')) :

            // return a admin_notice if not options page
            foreach ( $wp_functions as $function ) {
                if (!function_exists($function)) :
                    return $this->adminNotice(
                        __(
                            sprintf(
                                'It looks like that function %s doesn`t exists. 
                            Make sure you use the WpNonces Instance on a options/settings page',
                                $function
                            ),
                            'wpNonce'
                        )
                    );
                endif;
            }
        endif;

        $this->nonce = wp_create_nonce($this->action);
    }

    /**
     * This function fire a action-hook to admin notices
     * Default type is "error"
     *
     * @param string $message // Message to hook to admin_notices
     * @param string $type    // available types are
     *                        // ['success', 'info', 'warning', 'error']
     *
     * @return string
     * @see    https://codex.wordpress.org/Plugin_API/Action_Reference/admin_notices
     */
    protected function adminNotice(string $message, string $type = 'error')
    {
        $class = sprintf('notice notice-%s is-dismissible', $type);

        $out = sprintf(
            '<div class="%1$s"><p>%2$s</p></div>',
            esc_attr($class),
            esc_html($message)
        );

        return add_action('admin_notices', $out);
    }

    /**
     * Throw a Error-Message if not in WordPress
     *
     * @param string $exceptionClass // Error Class to throw
     * @param string $message        // Message of the Error
     *
     * @return string
     * @throws \Exception
     */
    protected function throwError($exceptionClass, $message)
    {
        throw new $exceptionClass($message);
    }

}
